packages/block-editor/src/components/link-control/index.js

```diff
@@ -138,11 +139,45 @@ function LinkControl( {
 
 	const [ settingsOpen, setSettingsOpen ] = useState( false );
 
-	const [ internalUrlInputValue, setInternalUrlInputValue ] =
-		useInternalInputValue( value?.url || '' );
+	const [ internalControlValue, setInternalControlValue ] =
+		useInternalValue( value );
 
-	const [ internalTextInputValue, setInternalTextInputValue ] =
-		useInternalInputValue( value?.title || '' );
+	const valueHasChanges =
+		value && ! isShallowEqualObjects( internalControlValue, value );
+
+	const setInternalURLInputValue = ( nextValue ) => {
+		setInternalControlValue( {
+			...internalControlValue,
+			url: nextValue,
+		} );
+	};
+
+	const setInternalTextInputValue = ( nextValue ) => {
+		setInternalControlValue( {
+			...internalControlValue,
+			title: nextValue,
+		} );
+	};
+
+	const setInternalSettingValue = ( nextValue ) => {
+		const settingsKeys = settings.map( ( { id } ) => id );
+
+		// Only apply settings values which are defined in the settings prop.
+		const settingsUpdates = Object.keys( nextValue ).reduce(
+			( acc, key ) => {
+				if ( settingsKeys.includes( key ) ) {
+					acc[ key ] = nextValue[ key ];
+				}
+				return acc;
+			},
+			{}
+		);
+
+		setInternalControlValue( {
+			...internalControlValue,
+			...settingsUpdates,
+		} );
+	};
```

Comment:
**Question**: would it be helpful if these methods were all internal to the `useInternalValue` hook?

Reply:
I agree that  it makes sense for all this logic to live in the hook.

Reply:
This would require the hook to know about `settings` 🤦

Reply:
I resolved this via partial application